### ISSUE TITLE
Disabling "Wiki Links > Enable links to missing tiddlers" Breaks Tags and Type:

### ISSUE THREAD

**Issue by GurjinderSingh** [Jermolene/TiddlyWiki5#2839]
Disabling "Enable links to missing tiddlers" Breaks Tags and Type.

Step to reproduce:
1. Download fresh TiddlyWiki from http://tiddlywiki.com
2. Go to : ControlPanel > settings (Tab) > Wiki Links ($:/core/ui/ControlPanel/Settings/MissingLinks)
3. uncheck "Enable links to missing tiddlers"  ($:/config/MissingLinks)
4. click create new tiddler
5. check type Drop down
6. add few tags save tiddler
7. create another tiddler check tags drop down. 

# type
![2017-04-23_230556](https://cloud.githubusercontent.com/assets/3470709/25315836/709cc0fc-2879-11e7-838c-eee7adbc9341.png)
# Tags
![2017-04-23_230910](https://cloud.githubusercontent.com/assets/3470709/25315860/e6724be4-2879-11e7-9ed0-99d685426df6.png)

recheck  Enable links to missing tiddlers now everything back to normal. 

![2017-04-23_231103](https://cloud.githubusercontent.com/assets/3470709/25315891/667d5dec-287a-11e7-87f7-107dcd20d225.png)

**Comment by pmario** [comment]
Can confirm the described behavior. ... It's a CSS problem. ... May be some templates have to be adjusted too, so we can assign proper styling for "disabled links"

**Comment by pmario** [comment]
Did you find any other side effects??

**Comment by GurjinderSingh** [comment]
in ControlPanel above that setting same css/html problem convert link to simple text. i think problem exists in shadows and system some ui related parts only that's all i found.

Editor Toolbar >Enable or disable the editor toolbar: "Show editor toolbar" 

thanks for fast reply and great work

**Comment by croulibri** [comment]
Thanks a lot for your discovery !

I was facing these two problems for something like a year in some of my TiddlyWiki without finding the origin of the problem.
In didn't find any other side effect of having selected "Enable links to missing tiddlers"

**Comment by GurjinderSingh** [comment]
i found problem in my wiki filled with thousands tiddlers using from 2013. was scared that all data need to copy paste thought my wiki got problem because i have updated it from too old version so its may be old data creating problem so downloaded new one copy pasted 5 tiddlers then got in to missing tiddlers wiki links so its time to do setting's boom new wiki got same problem then reversed some setting problem gone then did same on old :) saved my days. 

a feature request in old wiki when i save my wiki a version of wiki is saved in a backup directory. example

root_directory/mywiki.html
root_directory/backup_dir/
>> add tiddler > now save
updated wiki saved 
root_directory/mywiki.html

**Comment by GurjinderSingh** [comment]
Update: partially fixed tag are fine now. content type and field name list still broken.

**Comment by pmario** [comment]
@GurjinderSingh ... IMO this is a completely different problem. Please move your 2 posts to a new issue!

**Comment by tobibeer** [comment]
@Jermolene: 

The problem appears to be that the LinkWidget does not know that it is executed from within a LinkCatcherWidget.

In other words, [this line](https://github.com/Jermolene/TiddlyWiki5/blob/06ea4060cb44bf40ca266cc35f1318f5b85cc182/core/modules/widgets/link.js#L40) is problematic:

```
missingLinksEnabled = !(this.hideMissingLinks && this.isMissing && !this.isShadow);
```

It should also yield true for any `<$link/>` within a `<$linkcatcher/>`. So, in pseudo-code:

```
missingLinksEnabled = !(this.hideMissingLinks && this.isMissing && !this.isShadow) || this.withinLinkCatcher;
```

Right now, when a LinkWidget is rendered, I presume it does not verify the widget tree as to whether or not it is contained in a LinkCatcherWidget. So, the task would be: Figure out the simplest way to make a Linkwidget aware of being inside a LinkCatcher.

Alternatively, [this line|(https://github.com/Jermolene/TiddlyWiki5/blob/06ea4060cb44bf40ca266cc35f1318f5b85cc182/core/modules/widgets/link.js#L42):

```
if(useWikiLinks && missingLinksEnabled) {
```

...could equally be changed to 

```
if(useWikiLinks && missingLinksEnabled || this.withinLinkCatcher) {
```

In other words, when the LinkWidget is rendered, no tm-navigate is rendered yet... which is the hook that the LinkCatcherWidget exploits to re-route the click handler to do something else. However, we need a certain awareness in the LinkWidget by the time it is rendered.

**Comment by Jermolene** [comment]
Hi @tobibeer widgets interact with their parents via the `getVariable()` method. We try to avoid coupling the core widgets.

I think it was a mistake to be controlling the missing link behaviour via a configuration tiddler. We should instead be using a variable (like we do for `tc-wikilinks`), and then having a new wrapper in the page template that sets the variable from the currently defined configuration tiddler. Then the filter dropdown could be wrapped in a set widget that re-enables missing links.

**Comment by pmario** [comment]
I think it's just a CSS problem and can be easily fixed that way. ... BUT we have about 100 pending PRs... .. So I personally try to avoide new ones ... except translations.

**Comment by tobibeer** [comment]
@Jermolene, Concerning your proposal, I don't have an example off the top of my head where a setting translates to a global variable. Anything you could point me to? I might as well give it a shot to implement your proposed changes. Might be a few troublesome linkcatcher-links, but that's that then.

*However*, alternatively, the [LinkWidget](https://tiddlywiki.com/#LinkWidget) could also be extended to get a new property, e.g. `never-missing=Yes/No`, defaulting to 'no'. That way, we could instruct the affected links that they must never respect any *missingLinksEnabled* settings.

So, for example, [this line](https://github.com/Jermolene/TiddlyWiki5/blob/06ea4060cb44bf40ca266cc35f1318f5b85cc182/core/ui/EditTemplate/type.tid#L17) in [$:/core/ui/EditTemplate/type](https://github.com/Jermolene/TiddlyWiki5/blob/06ea4060cb44bf40ca266cc35f1318f5b85cc182/core/ui/EditTemplate/type.tid#L17) could then be changed to specify not only the `to` attribute but also the `never-missing` atribute as `yes`.

**Comment by Jermolene** [comment]
Hi @tobibeer 

> Concerning your proposal, I don't have an example off the top of my head where a setting translates to a global variable. Anything you could point me to?

You can see a bunch of similar examples in [$:/PageTemplate](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/ui/PageTemplate.tid)

**Comment by tobibeer** [comment]
Or even simplified from a `nevermissing=Yes/No` attribute, we could just use a `force` attribute w/o value so as to say "force this to always be rendered as a link".

**Comment by pmario** [comment]
@Jermolene -- Can be closed. There are no problems anymore.

**Comment by Jermolene** [comment]
Thank you @pmario
